src/bootstrap/suggest.rs

```diff
@@ -0,0 +1,142 @@
+use crate::builder::Builder;
+use glob::Pattern;
+use std::fmt::{self, Display};
+
+/// Suggests a list of possible `x.py` commands to run based on modified files in branch.
+pub fn suggest(build: &Builder<'_>, run: bool) {
+    // Find the commit the current branch split from master.
+    let merge_base = build
+        .config
+        .git()
+        .arg("merge-base")
+        .arg("HEAD")
+        .arg("origin/master")
+        .output()
+        .expect("failed to execute `git`");
+
+    // `git` shouldn't ever give us non-utf8 output.
+    let merge_base = String::from_utf8(merge_base.stdout).expect("`git` gave non-utf8 commit?!");
+
+    // Note that `merge_base` ends in a newline; it must be trimmed.
+    let modified_files = build
+        .config
+        .git()
+        .arg("diff")
+        .arg("--name-only")
+        .arg(merge_base.trim())
+        .output()
+        .expect("failed to execute `git`");
+
+    let modified_files =
+        String::from_utf8(modified_files.stdout).expect("`git` gave non-utf8 commit?!");
+
+    let modified_files: Vec<_> = modified_files.split_terminator('\n').collect();
+
+    println!("=== MODIFIED FILES ===");
+    for file in &modified_files {
+        println!("{}", file);
+    }
+    println!("======================\n");
+
+    let mut suggestions = Vec::new();
+    for sug in SUGGESTIONS {
+        let glob = Pattern::new(sug.0).expect("Found invalid glob pattern!");
+
+        for file in &modified_files {
+            if glob.matches(file) {
+                suggestions.extend_from_slice(sug.1);
+            }
+        }
+    }
+
+    suggestions.sort();
+    suggestions.dedup();
+
+    println!("=== SUGGESTIONS ===");
+    for sug in &suggestions {
+        println!("{}", sug);
+    }
+    println!("===================");
+
+    if run {
+        for _sug in suggestions {
+            // HELP!: how should I use `rustbuild`'s internal plumbing to achieve this?
```

Comment:
1. Add a new tool in `src/tools`; take a look at expand-yaml-anchors or something for ideas.
2. Add a new `Step` which knows how to build that tool and returns the path as its `Output`.
3. Use `builder.run(my_tool_path)`

Reply:
I'm not particularly keen on moving it into its own tool now. If the tool itself stays quite small and static but suggestions are dynamically loaded then it kind of seems unnecessarily complex but I'll move it into `src/tools` if you want.

This bit of code is really about actually running the suggestions, but that is addressed above? It seems pretty important to the whole tool for it to be able to run the suggested tests somehow (for CI or  whatever).